net.solarnetwork.node.weather.ibm.wc/src/net/solarnetwork/node/weather/ibm/wc/BasicWCClient.java

```diff
@@ -0,0 +1,246 @@
+
+import java.io.IOException;
+import java.math.BigDecimal;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Date;
+import java.util.List;
+import org.joda.time.LocalTime;
+import com.fasterxml.jackson.databind.JsonNode;
+import net.solarnetwork.node.domain.GeneralDayDatum;
+import net.solarnetwork.node.support.JsonHttpClientSupport;
+
+public class BasicWCClient extends JsonHttpClientSupport implements WCClient {
+
+	/** The default value for the {@code baseUrl} property. */
+	public static final String DEFAULT_BASE_URL = "https://api.weather.com";
+
+	/** The daily forecast url format */
+	public static final String DEFAULT_DAILY_FORECAST_URL = "v3/wx/forecast/daily/%s?icaoCode=%s"
+			+ "&language=%s&format=%s&units=%s&apiKey=%s";
+
+	/** The default output format */
+	public static final String DEFAULT_FORMAT = "json";
+
+	/** The default value for the {@code locationTemplate} property */
+	public static final String DEFAULT_LOCATION = "AU";
+
+	/** The hourly forecast url format */
+	public static final String DEFAULT_HOURLY_FORECAST_URL = "v3/wx/forecast/hourly/%s?icaoCode=%s"
+			+ "&language=%s&format=%s&units=%s&apiKey=%s";
+
+	/** The default value for the {@code apiKeyTemplate} property. */
+	public static final String DEFAULT_API_KEY = "abc123";
+
+	/** The default value for the {@code languageTemplate} property. */
+	public static final String DEFAULT_LANGUAGE = "en-US";
+
+	/** The default value for the {@code units} property. */
+	public static final String DEFAULT_UNITS = "e";
+
+	public static final boolean DEFAULT_COMPRESSION = true;
+
+	private String baseUrl;
+	private String dailyForecastUrl;
+	private String hourlyForecastUrl;
+	private String format;
+	private String language;
+	private String units;
+
+	public BasicWCClient() {
+		this.baseUrl = DEFAULT_BASE_URL;
+		this.dailyForecastUrl = DEFAULT_DAILY_FORECAST_URL;
+		this.hourlyForecastUrl = DEFAULT_HOURLY_FORECAST_URL;
+		this.format = DEFAULT_FORMAT;
+		this.language = DEFAULT_LANGUAGE;
+		this.units = DEFAULT_UNITS;
+
+		// set default compression
+		this.setCompress(DEFAULT_COMPRESSION);
+	}
+
+	private String getURL(String requestURL, String location, String apiKey, String datumPeriod) {
+
+		return this.baseUrl + "/"
+				+ String.format(requestURL, datumPeriod, location, language, format, units, apiKey);
+	}
+
+	/**
+	 * For the daily forecast, each part of the data (ie. max temp, min temp,
+	 * narrative) is an array, with each entry corresponding with a day of the
+	 * week. The more detailed features in "daypart" such as cloud cover are
+	 * split into day and night, meaning two entries for each day.
+	 */
+	private BigDecimal parseBigDecimal(JsonNode n) {
+		if ( n.isNull() ) {
+			return null;
+		}
+		return new BigDecimal(n.asLong());
+	}
+
+	private LocalTime parseTimeValue(JsonNode n) {
+		if ( n == null || n.isNull() ) {
+			return null;
+		}
+		return new LocalTime(n.asLong() * 1000);
```

Comment:
Does this handle time zones correctly? When I read this, it says to me "get this absolute time as a local time in the time zone the JVM is running in." What we really want is "get this absolute time as a local time in the time zone the data is in."

I think to work correctly, you may have to pass the time zone of the location the data is from to the constructor. For example a node might be running in the `Pacific/Auckland` time zone but collecting weather data for the `America/New_York` time zone. Add a unit test for that scenario, and verify you end up with a sunset/sunrise time that is appropriate still.

Where can you get the location's time zone from? I notice there is a `validTimeLocal` property in the response that includes an appropriate local time zone. You should be able to use the time zone included there to correctly generate `LocalTime` and `LocalDate` values. Then you could pass the `DateTimeZone` found there into this method could be updated to include a `DateTimeZone` argument. 

**Note** that the time zone offset might change when dates go in/out of DST... so realistically you probably need to parse and use the same `validTimeLocal` array index value that is the same array index value of the node being parsed here... that is if you're parsing index 1 of `sunriseTimeUtc` you need to use index 1 of `validTimeLocal` to find the time zone.

Reply:
The change I made to fix this is to include the GMT offset in the SimpleTimeFormat string - is this correct?

Reply:
Ah I didn't notice the `moonriseTimeLocal` values that are in the daily data. So those are great, as long as you parse them into `DateTime` objects using Joda Time's `DateTimeFormatter`, not SimpleDateFormat like you're doing here because you're still not passing the time zone into the `LocalTime` constructor.

In the `7day_daily.json` file the 2nd `sunriseTimeLocal` value is "2018-07-06T05:37:04-0600" which you are verifying in the `WCClientTests.readDailyForecast` method. I'd expect your test to verify that the final string value that gets set into your datum object is `05:37` however your unit test verifies this as `23:37` which doesn't make sense for the sun to rise at 11:37pm local time, right?

I suggest you use Joda's `org.joda.time.format.ISODateTimeFormat.dateTimeParser()` to get a `DateTimeFormatter` suitable for parsing the `*TimeLocal` values in JSON into `DateTime` objects (which should preserve the time zone), which you can then easily convert to `LocalTime` with the `toLocalTime() method.